pydantic/schema.py

```diff
@@ -723,42 +748,79 @@ def encode_default(dft: Any) -> Any:
-        ):
-            # Is numeric type
-            attrs = ('gt', 'lt', 'ge', 'le', 'multiple_of')
-            numeric_type = next(t for t in numeric_types if issubclass(annotation, t))  # pragma: no branch
-            constraint_func = _map_types_constraint[numeric_type]
-        elif issubclass(annotation, ConstrainedList):
-            attrs = ('min_items', 'max_items')
-            constraint_func = conlist
-        if attrs:
-            kwargs = {
-                attr_name: attr
-                for attr_name, attr in ((attr_name, getattr(field_info, attr_name)) for attr_name in attrs)
-                if attr is not None
-            }
-            if kwargs:
-                constraint_func = cast(Callable[..., type], constraint_func)
-                return constraint_func(**kwargs)
+    item_type = get_item_type(annotation)
+    if not isinstance(item_type, type):
+        return annotation
+
+    attrs: Tuple[str, ...] = ()
+    constraint_func: Optional[Callable[..., type]] = None
+    if lenient_issubclass(annotation, ConstrainedList):
+        attrs = ('min_items', 'max_items')
+        constraint_func = conlist
+    elif issubclass(item_type, str) and not issubclass(item_type, (EmailStr, AnyUrl, ConstrainedStr)):
+        attrs = ('max_length', 'min_length', 'regex')
+        constraint_func = constr
+    elif issubclass(item_type, bytes):
+        attrs = ('max_length', 'min_length', 'regex')
+        constraint_func = conbytes
+    elif issubclass(item_type, numeric_types) and not issubclass(
+        item_type, (ConstrainedInt, ConstrainedFloat, ConstrainedDecimal, ConstrainedList, bool)
+    ):
+        # Is numeric type
+        attrs = ('gt', 'lt', 'ge', 'le', 'multiple_of')
+        numeric_type = next(t for t in numeric_types if issubclass(item_type, t))  # pragma: no branch
+        constraint_func = _map_types_constraint[numeric_type]
+
+    unused_constraints = [
+        f for f in validation_attribute_to_schema_keyword if f not in attrs and getattr(field_info, f) is not None
+    ]
+    if unused_constraints:
+        raise ValueError(
+            f'{field_name}: the following field constraints are set but not enforced: {", ".join(unused_constraints)}'
+        )
+
+    if attrs:
+        kwargs = {
+            attr_name: attr
+            for attr_name, attr in ((attr_name, getattr(field_info, attr_name)) for attr_name in attrs)
+            if attr is not None
+        }
+        if kwargs:
+            constraint_func = cast(Callable[..., type], constraint_func)
+            return constraint_func(**kwargs)
     return annotation
 
 
+def get_item_type(annotation: Any) -> Any:
+    if is_literal_type(annotation):
+        return None
+    origin = getattr(annotation, '__origin__', None)
+    if origin is None:
+        return annotation
+
+    args: Tuple[Any, ...] = annotation.__args__
+
+    if origin is Union:
+        if sum(a is not NoneType for a in args) == 1:  # type: ignore
```

Comment:
This feels kind of fragile, for example, what if you had `Optional[Union[str, bytes]]`? It seems like there is no reason the `get_annotation_from_field_info` function shouldn't be able to handle that, but this would reject it.

I'm wondering if maybe there is a way to refactor this pair of functions so that it is applied on a per-union-member basis?

Reply:
same as above.